solara/components/input.py

```diff
@@ -276,113 +288,70 @@ def Page():
     * `continuous_update`: Whether to call the `on_value` callback on every change or only when the input loses focus or the enter key is pressed.
     * `clearable`: Whether the input can be cleared.
     """
+
+    def str_to_int(value: Optional[str]):
+        if value is not None and value != "":
+            try:
+                return int(value)
+            except ValueError:
+                raise ValueError("Value must be an integer")
+        else:
+            if optional:
+                return None
+            else:
+                raise ValueError("Value cannot be empty")
+
     return _InputNumeric(
-        int,
+        str_to_int,
         label=label,
         value=value,
         on_value=on_value,
         disabled=disabled,
-        optional=optional,
         continuous_update=continuous_update,
         clearable=clearable,
     )
 
 
-def _use_input_numeric(
-    str_to_numeric: Callable[[str], T],
-    value: Union[None, T, solara.Reactive[Optional[T]], solara.Reactive[T]],
+def _use_input_type(
+    input_value: Union[None, T, solara.Reactive[Optional[T]], solara.Reactive[T]],
+    parse: Callable[[Optional[str]], T],
+    stringify: Callable[[Optional[T]], str],
     on_value: Union[None, Callable[[Optional[T]], None], Callable[[T], None]] = None,
-    optional: bool = False,
 ):
-    # TODO: make this more type safe
-    reactive_value = solara.use_reactive(value, on_value)  # type: ignore
-    del value, on_value
+    reactive_value = solara.use_reactive(input_value, on_value)  # type: ignore
+    del input_value, on_value
+    string_value, set_string_value = solara.use_state(stringify(reactive_value.value) if reactive_value.value is not None else None)
+    # Workaround to hook() getting a stale string_value
```

Comment:
```suggestion
    # Use a ref to make sure sync_back_input_value() does not get a stale string_value
```
it's not a workaround I would say, but it's more like a common pattern in react